### PR TITLE
Breaking: Refactor and streamline hint-compat-api

### PR DESCRIPTION
**Pull request by antross**
## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- [x] Added/Updated related documentation.
- [x] Added/Updated related tests.

## Short description of the change(s)

Posting as a draft due to a few remaining tasks, but feel free to review.

Tasks:
* [x] Restore support for user-provided ignore lists
* [x] Resolve multiple prefixed properties in the same block
* [x] Combine multiple versions of the same browser in reports
* [x] Pass relevant portions of the existing test suite
* [x] Trim test suite to focus on higher-level behavior (lower-level covered by helper unit tests)
* [x] Update documentation

### CODE REVIEW

File: packages/hint-compat-api/src/utils/parse-supports.ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+ *         }
+ *     ]
+ * }
+ * ```
+ */
+export const parseSupports = (params: string): DeclarationGroup | null => {
+    try {
+        return parseParams(params);
+    } catch (e) {
+        return null;
```
Comment (antross) [inline packages/hint-compat-api/src/utils/parse-supports.ts]: Looking at this again I'm thinking I should throw a specific type of error for known invalid states and only filter errors of that type out here. This would avoid silencing unexpected bugs is the parser...

File: packages/hint-compat-api/src/utils/ignore.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+/**
+ * Determine the set of ignored features, given the provided defaults
+ * and user-specified `ignore` and/or `enable` options.
+ */
+export const resolveIgnore = (options: any, defaults: string[]) => {
+    const ignore = new Set(defaults);
+
+    if (options && options.ignore) {
+        options.ignore.forEach((entry: string) => {
```
Comment (antross) [inline packages/hint-compat-api/src/utils/ignore.ts]: Not sure why I wrote this as a `forEach`. We've been preferring `for..of` loops lately.

File: packages/hint-compat-api/src/utils/ignore.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+            ignore.add(entry);
+        });
+    }
+
+    if (options && options.enable) {
+        options.enable.forEach((entry: string) => {
```
Comment (antross) [inline packages/hint-compat-api/src/utils/ignore.ts]: `for..of`

File: packages/hint-compat-api/src/utils/filter-supports.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    if (!root) {
+        return null;
+    }
+
+    try {
+        const supported = filterItem(root, hasAtSupports);
+
+        return supported.length ? supported : null;
+    } catch (e) {
+        return null;
```
Comment (antross) [inline packages/hint-compat-api/src/utils/filter-supports.ts]: Similar to the parser, I should throw and catch a known error type here. Then other errors can be re-thrown to avoid silencing unexpected issues.

File: packages/hint-compat-api/src/utils/browsers.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    const results: string[] = [];
+
+    // Sort and serialize version ranges for each browser name.
+    for (const [name, versions] of versionsByName) {
+        versions.sort((v1, v2) => {
+            return semver.compare(coerce(v1), coerce(v2));
+        });
+
+        const ranges = [];
```
Comment (antross) [inline packages/hint-compat-api/src/utils/browsers.ts]: I'm surprised TypeScript didn't complain about an implicit `any` here...
```suggestion
        const ranges: string[] = [];
```

File: packages/hint-compat-api/src/css.ts [hunk 6]
```diff
@@ -1,29 +1,227 @@
...
+    if (unsupported) {
+        context.report({ feature: `${node.selector}`, node, unsupported });
+    }
+
+    context.walk(node, context);
+};
+
+/**
+ * Filter reports for the same CSS property (prefixed or not) and only emit
+ * those for browsers that were unsupported in ALL reports (if any), favoring
```
Comment (antross) [inline packages/hint-compat-api/src/css.ts]:
```suggestion
 * those for browsers that were unsupported in ALL reports in the same block (if any), favoring
```

File: packages/hint-compat-api/docs/css.md [hunk 7]
```diff
@@ -1,84 +1,68 @@
-# Deprecated CSS features (`compat-api/css`)
+# Supported CSS features (`compat-api/css`)
 
 ## What does the hint check?
 
-`compat-api/css` checks if the CSS features used are deprecated in the
-[targeted browsers][browser-context].
+`compat-api/css-next` checks if the CSS features used are
```
Comment (molant) [inline packages/hint-compat-api/docs/css.md]: Files are `css.(ts|md)`, this should be consistent.

```suggestion
`compat-api/css` checks if the CSS features used are
```

File: packages/hint-compat-api/docs/css.md [hunk 8]
```diff
@@ -1,84 +1,68 @@
...
+and feature detection via `@supports` blocks.
 
 ### Examples that **trigger** the hint
 
-The [box-lines][box-lines] property
-was added with the `-webkit-` prefix for Chrome and removed from versions of
-Chrome 67 and onwards.
-Targeted Chrome browsers of versions 67 and up will trigger the hint.
+The [box-flex][box-flex] was
+never added for any version of Internet Explorer. If Internet Explorer is
```
Comment (molant) [inline packages/hint-compat-api/docs/css.md]: Should it be `added to`? I think there's another one later on in case it has to be changed. Also later in the HTML one you use `was never added on any version`.
 
```suggestion
never added to any version of Internet Explorer. If Internet Explorer is
```

File: packages/hint-compat-api/docs/css.md [hunk 9]
```diff
@@ -1,84 +1,68 @@
...
-    }
+    display: grid;
 }
 ```
 
 ### Examples that **pass** the hint
 
-The [background][background] property was never
-removed for any browser. It should always pass the hint.
+The [charset][charset]
```
Comment (molant) [inline packages/hint-compat-api/docs/css.md]: I think we should add an `@supports` example in this section.

File: packages/hint-compat-api/docs/css.md [hunk 10]
```diff
@@ -101,7 +95,7 @@ In the `.hintrc` file:
     "connector": {...},
     "formatters": [...],
     "hints": {
-        "compat-api/css": ["error", {
+        "compat-api/css-next": ["error", {
```
Comment (molant) [inline packages/hint-compat-api/docs/css.md]:
```suggestion
        "compat-api/css": ["error", {
```

File: packages/hint-compat-api/docs/css.md [hunk 11]
```diff
@@ -120,7 +114,7 @@ In the `.hintrc` file:
...
```
Comment (molant) [inline packages/hint-compat-api/docs/css.md]:
```suggestion
        "compat-api/css": ["error", {
```

File: packages/hint-compat-api/docs/css.md [hunk 12]
```diff
@@ -1,84 +1,68 @@
...
+does not have any information about when it was added to various browsers
+according to the  [MDN compat data][browser-compat]. No error is thrown
+if there is no information available on when, or if, the feature was added.
 
 ```css
 .example {
-    box-lines: single;
+    text-transform: capitalize;
 }
 ```
```
Comment (molant) [inline packages/hint-compat-api/docs/css.md]: We talked about listing the current limitations of the hint too.

File: packages/hint-compat-api/docs/html.md [hunk 13]
```diff
@@ -1,75 +1,73 @@
-# Deprecated HTML features (`compat-api/html`)
+# Supported HTML features (`compat-api/html`)
 
 ## What does the hint check?
 
-`compat-api/html` checks if the HTML elements and attributes used are
-deprecated in the [targeted browsers][browser-context].
+`compat-api/html-next` checks if the HTML elements and attributes used are
```
Comment (molant) [inline packages/hint-compat-api/docs/html.md]:
```suggestion
`compat-api/html` checks if the HTML elements and attributes used are
```

File: packages/hint-compat-api/docs/html.md [hunk 14]
```diff
@@ -92,7 +98,7 @@ In the `.hintrc` file:
     "connector": {...},
     "formatters": [...],
     "hints": {
-        "compat-api/html": ["error", {
+        "compat-api/html-next": ["error", {
```
Comment (molant) [inline packages/hint-compat-api/docs/html.md]:
```suggestion
        "compat-api/html": ["error", {
```

File: packages/hint-compat-api/docs/html.md [hunk 15]
```diff
@@ -111,7 +117,7 @@ In the `.hintrc` file:
...
```
Comment (molant) [inline packages/hint-compat-api/docs/html.md]:
```suggestion
        "compat-api/html": ["error", {
```

File: packages/hint-compat-api/docs/css.md (hunk 8 above)
Comment (antross) [reply packages/hint-compat-api/docs/css.md]: Sounds better to me - most of this section was pre-existing. I just moved things around a bit.

File: packages/hint-compat-api/src/css.ts [hunk 16]
```diff
@@ -1,29 +1,227 @@
...
+
+    if (reportData) {
+        if (!reports.has(prop)) {
+            reports.set(prop, []);
+        }
+
+        // Decl had a report, add it to the list to resolve later.
+        reports.get(prop)!.push(reportData);
+    } else {
+        // Decl was fully supported; no need to test other decls in this block.
```
Comment (molant) [inline packages/hint-compat-api/src/css.ts]: It took me a while to understand what was going on here with this comment. Maybe something like the following?

> Decl was fully supported; no need to test other decls for same property in this block so setting to `null` to `return` at the beginning of the function.

File: packages/hint-compat-api/tests/html.ts [hunk 17]
```diff
@@ -15,231 +13,80 @@ const generateHTMLConfig = (fileName: string) => {
...
+        },
+        {
+            name: 'Does not report manually ignored HTML features',
+            serverConfig: generateHTMLConfig('values')
+        }
+    ],
+    {
+        browserslist: targetBrowsers,
+        hintOptions: { enable: ['integrity'], ignore: ['input[type=color]'] },
+        parsers: ['css']
```
Comment (antross) [inline packages/hint-compat-api/tests/html.ts]:
```suggestion
```

File: packages/hint-compat-api/tests/html.ts [hunk 18]
```diff
@@ -15,231 +13,80 @@ const generateHTMLConfig = (fileName: string) => {
...
+                    message: 'input[type=color] is not supported by ie 10-11.',
+                    position: { match: 'input type="color"' }
+                }
+            ],
+            serverConfig: generateHTMLConfig('values')
+        }
+    ],
+    {
+        browserslist: targetBrowsers,
+        parsers: ['css']
```
Comment (antross) [inline packages/hint-compat-api/tests/html.ts]:
```suggestion
```

File: packages/hint-compat-api/src/utils/browsers.ts [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+
+        // Outer loop runs once per start of each range.
+        for (let i = 0; i < versions.length; i++) {
+            const start = versions[i];
+
+            // Inner loop jumps to the end of a range.
+            while (parseInt(versions[i + 1]) - parseInt(versions[i]) <= 1) {
+                i++;
```
Comment (antross) [inline packages/hint-compat-api/src/utils/browsers.ts]: @sarvaje You wanted to think about a different approach to doing this without modifying `i` inside the loop.

File: packages/hint-compat-api/src/css.ts [hunk 20]
```diff
@@ -1,29 +1,227 @@
...
+    } else {
+        validateDeclValue(node, context);
+    }
+};
+
+const validateDeclInBlock = (node: Declaration, reports: ReportMap, context: Context): void => {
+    const prop = vendor.unprefixed(node.prop);
+
+    // If any matching decl was already fully supported, no need to evaluate further.
+    if (reports.get(prop) === null) {
```
Comment (antross) [inline packages/hint-compat-api/src/css.ts]: Use a constant or enum with a name like `Passed` instead of `null` for clarity.

File: packages/hint-compat-api/src/utils/browsers.ts (hunk 19 above)
Comment (antross) [reply packages/hint-compat-api/src/utils/browsers.ts]: @sarvaje Here's what I ended up changing it to:

```ts
for (let i = 0, start = versions[0]; i < versions.length; i++) {
    if (parseInt(versions[i + 1]) - parseInt(versions[i]) <= 1) {
        continue; // Continue until the end of a range.
    }

    // Format current range as either `start` or `start-end`.
    if (start === versions[i]) {
        ranges.push(start);
    } else {
        ranges.push(`${start}-${versions[i]}`);
    }

    // Remember the start of the next range.
    start = versions[i + 1];
}
```

Comment (sarvaje) [reply packages/hint-compat-api/src/utils/browsers.ts]: I like this version more than the previous one hehehe.

File: packages/hint-compat-api/src/utils/filter-supports.ts [hunk 21]
```diff
@@ -0,0 +1,78 @@
+import difference = require('lodash/difference');
+import intersection = require('lodash/intersection');
+import union = require('lodash/union');
+
+import { getSupported } from '@hint/utils/dist/src/compat/support';
+
+import { parseSupports, Declaration, DeclarationGroup } from './parse-supports';
+
+class InvalidInputError extends Error {
```
Comment (sarvaje) [inline packages/hint-compat-api/src/utils/filter-supports.ts]: 👍

File: packages/hint-compat-api/src/utils/parse-supports.ts [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+};
+
+export type DeclarationGroup = {
+    type: 'and' | 'or' | 'not';
+    nodes: (Declaration | DeclarationGroup)[];
+};
+
+type Token = '(' | ')' | 'not' | 'and' | 'or' | Declaration;
+
+class InvalidInputError extends Error {
```
Comment (sarvaje) [inline packages/hint-compat-api/src/utils/parse-supports.ts]: You have the same `InvalidInputError` in the file ` packages/hint-compat-api/src/utils/filter-supports.ts`, maybe you can move this to its own file and share it.

Comment (antross) [reply packages/hint-compat-api/src/utils/parse-supports.ts]: By keeping them distinct classes, if `parse-supports.ts` accidentally lets one escape, it won't get trapped by `filter-supports.ts` so we can see the issue and fix it.

I can give them unique names if we want, but I don't think we should share a single definition as each component should own it's own "ignored" error type.

Comment (antross) [reply packages/hint-compat-api/src/utils/parse-supports.ts]: Alternatively I could just add comment clarifying why the error type is not shared if that works for you.

Comment (sarvaje) [reply packages/hint-compat-api/src/utils/parse-supports.ts]: hehehe I saw the same name/code and I thought it was the same.
Maybe a comment for future refactors would be nice :)
